test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,516 @@
+
+    wal_delay_end(g.server_1)
+    fiber_join(g.server_1, f)
+    wait_sync(g.cluster.servers)
+
+    box_cfg_update({g.server_1}, {election_mode = 'off'})
+    demote(g.server_1)
+end
+
+-- Promoting and demoting should work when everything is ok.
+g_common.test_ok = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting succeeds with elections off: %s', err))
+
+    wait_sync(g.cluster.servers)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'Demoting succeeds with elections off: %s', err))
+
+    box_cfg_update({g.server_1}, {election_mode = 'manual'})
+
+    wait_sync(g.cluster.servers)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'Promoting succeeds with elections on: %s', err))
+
+    wait_sync(g.cluster.servers)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'Demoting succeeds with elections on: %s', err))
+end
+
+-- Simultaneous promoting/demoting should fail.
+g_common.test_simultaneous = function(g)
+    wal_delay_start(g.server_1)
+
+    local term = g.server_1:election_term()
+    local f = promote_start(g.server_1)
+    g.server_1:wait_election_term(term + 1)
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'Simultaneous promote fails')
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'Simultaneous demote fails')
+
+    wal_delay_end(g.server_1)
+    fiber_join(g.server_1, f)
+    wait_sync(g.cluster.servers)
+    demote(g.server_1)
+end
+
+-- Promoting voter should fail.
+g_common.test_voter_promote = function(g)
+    box_cfg_update({g.server_1}, {election_mode = 'voter'})
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'Promoting voter fails')
+end
+
+-- Promoting should fail if it is interrupted from another server
+-- while writing new term to wal.
+g_common.test_wal_interfering_promote = function(g)
+    wal_delay_start(g.server_1)
+    local term = g.server_1:election_term()
+    local f = promote_start(g.server_1)
+    g.server_1:wait_election_term(term + 1)
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RAFT_WAIT_TERM_PERSISTED_DELAY', true)
+    end)
+    promote(g.server_2)
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_term(g.server_2:synchro_queue_term())
```

Comment:
I don't think you need an injection here. You can wait for `box.info.election.leader == g.server_2:instance_id()` after `promote(g.server_2)` and it will have the same effect. Same with the demote test.